### PR TITLE
feat: weather data analysis

### PR DESCRIPTION
**Pull request by NasimFatima**

### CODE REVIEW

File: weatherman/weatherman.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+     * @param {String} directoryPath Path where files are stored.
+     * @param {String} date Date in form of year or Y/m.
+     * @param {boolean} month
+     * @param {string} calculateTask
+     */
+    constructor(directoryPath, date, month=False, calculateTask) {
+        this.directoryPath = directoryPath
+        this.date = date
+        this.withMonth = month
+        this.calculateTask = calculateTask
```
Comment (mareenanazir) [inline weatherman/weatherman.js]: we should rename this variable to something better

File: weatherman/index.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    alias: "d",
+    description: "date to run for",
+    type: "string",
+  })
+  .help()
+  .alias("help", "h").argv;
+
+if (argv.task === "-e" || argv.task === "-a") {
+  if (argv.date) {
+    const yearMonthFormat = validateYearAndMonth(argv.date);
```
Comment (mareenanazir) [inline weatherman/index.js]: we don't need to calculate this

File: weatherman/index.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    }
+  }
+}
+else {
+    console.log("ENTER VALID INPUT OR PRESS --help")
+}
+
+/**
+ * Join Current path with directory name
+ * @param {string} directoryName
```
Comment (mareenanazir) [inline weatherman/index.js]: follow the same doctstring convention throughout the code.

File: weatherman/utils/dateUtils.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable require-jsdoc */
+
+
+export const getShortMonthFromString = (dateString) => {
```
Comment (mareenanazir) [inline weatherman/utils/dateUtils.js]: why do we need to get this short month?

File: weatherman/utils/mathUtils.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/* eslint-disable require-jsdoc */
+
+export const getMinMaxObj = (objArray, property, Max=true ) => {
```
Comment (mareenanazir) [inline weatherman/utils/mathUtils.js]: we should make them separate functions

File: weatherman/utils/mathUtils.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    const average = (data.reduce((sum, elem) => sum + elem[property], 0) / data.length) * 100
+    const closest = data.reduce((a, b) => 
+        Math.abs(average - a[property]) <= Math.abs(average - b[property])
+            ? a
+            : b
+    );
+    return closest
+}
+
+function removeEmptyValue(data, property) {
```
Comment (mareenanazir) [inline weatherman/utils/mathUtils.js]: we should remove the empty values when reading the file

File: weatherman/fileReader.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  /**
+   * filter directory by given format
+   * @return {matchingFileNames}
+   */
+  readFileNamesByFormat() {
+    let matchingFileNames = [];
+    try {
+      matchingFileNames = fs
+        .readdirSync(this.directory)
+        .filter((file) => file.includes(this.format) === true);
```
Comment (mareenanazir) [inline weatherman/fileReader.js]: do we need to explicitly check for true?

File: weatherman/fileReader.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  /**
+   * Read Content of filtered files
+   * @return {weatherDataMap}
+   */
+  readFileContent() {
+    let weatherObj = {};
+    const weatherDataMap = [];
+    const matchingFileNames = this.readFileNamesByFormat();
+
+    if (matchingFileNames.length < 1) {
```
Comment (mareenanazir) [inline weatherman/fileReader.js]: can we remove this if from here?

File: weatherman/fileReader.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      return weatherDataMap;
+    }
+
+    matchingFileNames.forEach((file) => {
+      const data = fs.readFileSync(this.directory + file, "utf-8");
+      const fileData = d3.csvParseRows(data);
+      for (let j = 1; j < fileData.length; j++) {
+        weatherObj = Object.keys(WEATHER_FILE_HEADERS).reduce(
```
Comment (mareenanazir) [inline weatherman/fileReader.js]: clean the data here

Comment (mareenanazir) [reply weatherman/fileReader.js]: store only the fields that are required

File: weatherman/constants.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+
+export const WEATHER_FILE_HEADERS = {
```
Comment (mareenanazir) [inline weatherman/constants.js]: we dont need the headers that are not being used

File: weatherman/index.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  })
+  .option("date", {
+    alias: "d",
+    description: "date to run for",
+    type: "string",
+  })
+  .help()
+  .alias("help", "h").argv;
+
+if (argv.task === "-e" || argv.task === "-a") {
```
Comment (mareenanazir) [inline weatherman/index.js]: we should make this capable enough to handle two inputs at a time

File: weatherman/weatherman.js [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+import { FilReader } from "./fileReader.js"
+
+/**
+ * WeatherMan class .
+ */
+export class WeatherMan {
+
+    /**
+     * Function Constructor.
+     * @param {String} directoryPath Path where files are stored.
```
Comment (mareenanazir) [inline weatherman/weatherman.js]: we should add a hyphen after the variable name so the description does not get mixed with the variable name

File: weatherman/weatherman.js [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        }
+        return this.formatToMatchFiles
+    }
+
+    /**
+     * Read Files and calculate weather Stats
+     */
+    getWeatherStats() {
+        this.weatherStats = {
+            'Max' : {'MaxTemperatureC' : {}, 'MinTemperatureC' : {}, 'MaxHumidity' : {}},
```
Comment (mareenanazir) [inline weatherman/weatherman.js]: can we do this in some other way?

File: weatherman/weatherman.js [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        console.log(`LOWEST: ${stats.MaxTemperatureC.MaxTemperatureC}C on  ${formattedDateForOutput(stats.MinTemperatureC.Date)}`)
+        console.log(`HIGHEST HUMIDITY: ${stats.MaxHumidity.MaxHumidity}% on  ${formattedDateForOutput(stats.MaxHumidity.Date)}`)
+        console.log("****************************************")
+    }
+
+    /**
+     * Print Average MaxTemperature, Average Lowest Temperature and
+     * Average Mean Humidity in given format 
+     */
+    printAvgTempAndHumidity() {
```
Comment (mareenanazir) [inline weatherman/weatherman.js]: function name should be something more specific to the work being done in this

File: weatherman/index.js [hunk 15]
```diff
@@ -0,0 +1,54 @@
+import { createRequire } from "module";
+import { validateYearAndMonth, validateYear } from "./utils/dateUtils.js";
+import { WeatherMan } from "./weatherman.js";
+const require = createRequire(import.meta.url);
+const path = require("path");
```
Comment (mareenanazir) [inline weatherman/index.js]: why are we using imports and require both?
can't we use imports only?

**Review by mareenanazir — CHANGES_REQUESTED** [review]

File: weatherman/utils/dateUtils.js (hunk 4 above)
Comment (NasimFatima) [reply weatherman/utils/dateUtils.js]: short for compare input with file names, long for to display output

File: weatherman/utils/mathUtils.js (hunk 6 above)
Comment (NasimFatima) [reply weatherman/utils/mathUtils.js]: Resolved as discussed

File: weatherman/constants.js (hunk 10 above)
Comment (NasimFatima) [reply weatherman/constants.js]: these headers are used for index mapping

Comment (mareenanazir) [reply weatherman/constants.js]: we can only store the ones that we need to calculate, this is extra

File: weatherman/weatherman.js [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+            if (month)
+                this.formatToMatchFiles += `_${month}`
+            else 
+                this.formatToMatchFiles = ''
+        }
+    }
+
+    /**
+     * Read Files and calculate weather Stats
+     * @param { string } columns -Get data of given columns
```
Comment (mareenanazir) [inline weatherman/weatherman.js]: missing space after hyphen

File: weatherman/utils/mathUtils.js [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    if (data.length > 0) {
+        return data.reduce((min, elem) => Number(min[property]) < Number(elem[property]) ? min : elem)
+    }
+    return data   
+}
+
+export const getAvgObj = (objArray, property) => {
+    const data = removeEmptyValue(objArray, property)
+    const average = (data.reduce((sum, elem) => sum + elem[property], 0) / data.length) * 100
+    const closest = data.reduce((a, b) =>
```
Comment (mareenanazir) [inline weatherman/utils/mathUtils.js]: why are we calculating this closest?

File: weatherman/utils/mathUtils.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
+/* eslint-disable require-jsdoc */
+
+export const getMaxObj = (objArray, property) => {
+    const data = removeEmptyValue(objArray, property)
+    if (data.length > 0) {
```
Comment (mareenanazir) [inline weatherman/utils/mathUtils.js]: can we remove this if from here?

File: weatherman/utils/mathUtils.js [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+        return data.reduce((max, elem) => Number(max[property]) > Number(elem[property]) ? max : elem)
+    }
+    return data   
+}
+
+export const getMinObj = (objArray, property) => {
+    const data = removeEmptyValue(objArray, property)
+    if (data.length > 0) {
```
Comment (mareenanazir) [inline weatherman/utils/mathUtils.js]: try doing this without if

File: weatherman/index.js [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+  .option("date2", {
+    alias: "d2",
+    description: "date to run for e.g. 2004/03",
+    type: "string",
+  })
+  .help()
+  .alias("help", "h").argv;
+
+const directoryPath = buildPath("weatherfiles");
+if (argv.date1) {
```
Comment (mareenanazir) [inline weatherman/index.js]: can we use some better variable name?

File: weatherman/weatherman.js [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+        this.directoryPath = directoryPath
+        this.date = date
+    } 
+
+     /**
+     * Parse date string 
+     */
+    parseDate() {
+        this.formatToMatchFiles = getYear(this.date)
+        if (this.withMonth) {
```
Comment (mareenanazir) [inline weatherman/weatherman.js]: is `withMonth` an attribute of WeatherMan class?

File: weatherman/weatherman.js [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+            }
+        });
+        this.printHighLowTempAndHumidity()
+    }
+
+    /**
+     * Claculate Average MaxTemperature, Average Lowest Temperature and
+     * Average Mean Humidity 
+     */
+    getAvgTempAndHumidity() {
```
Comment (mareenanazir) [inline weatherman/weatherman.js]: we should rename the function to something better

File: weatherman/utils/dateUtils.js [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+        console.error(`EXCEPTION IN DATE PARSING ${err}`)
+    }
+
+}
+
+export const getYear = dateString => {
+    return parseDate(dateString).getFullYear()
+}
+
+export const formattedDateForOutput = dateString => {
```
Comment (mareenanazir) [inline weatherman/utils/dateUtils.js]: function name should be something more appropriate

**Review by mareenanazir — CHANGES_REQUESTED** [review]
Lets try to handle the dynamic path for the weather files as well.

File: weatherman/weatherman.js (hunk 21 above)
Comment (NasimFatima) [reply weatherman/weatherman.js]: flag set in calculateMonthStats function

File: weatherman/constants.js (hunk 10 above)
Comment (NasimFatima) [reply weatherman/constants.js]: resolved

File: weatherman/index.js [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+  .help()
+  .alias("help", "h").argv;
+
+let directoryPath;
+
+if (argv.path) directoryPath = argv.path;
+else directoryPath = buildPath("weatherfiles");
+
+if (argv.year) {
+  if (argv.task1 === "-e") {
```
Comment (mareenanazir) [inline weatherman/index.js]: what is task1 is not equal to e?

File: weatherman/index.js [hunk 25]
```diff
@@ -0,0 +1,65 @@
+import { createRequire } from "module";
+import { WeatherMan } from "./weatherman.js";
+const require = createRequire(import.meta.url);
+import * as path from "path";
+const yargs = require("yargs");
+
+const argv = yargs
+  .option("task1", {
```
Comment (mareenanazir) [inline weatherman/index.js]: change the names to something else task1 task2 are not really good names

File: weatherman/fileReader.js [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+  constructor(directory, format) {
+    this.directory = directory;
+    this.format = format;
+  }
+
+  /**
+   * filter directory by given format
+   * @return { matchingFileNames }
+   */
+  readFileNamesByFormat() {
```
Comment (mareenanazir) [inline weatherman/fileReader.js]: lets make these arrow functions

File: weatherman/index.js [hunk 27]
```diff
@@ -0,0 +1,65 @@
+import { createRequire } from "module";
+import { WeatherMan } from "./weatherman.js";
+const require = createRequire(import.meta.url);
+import * as path from "path";
+const yargs = require("yargs");
```
Comment (mareenanazir) [inline weatherman/index.js:5]: why both import and require

File: weatherman/index.js [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+    alias: "y",
+    description: "date to run for e.g. 2004",
+    type: "string",
+  })
+  .option("task2", {
+    alias: "t2",
+    description: "Task to run: -a",
+    type: "string",
+  })
+  .option("yearmonth", {
```
Comment (mareenanazir) [inline weatherman/index.js]: can we do something better than yearmonth?

File: weatherman/utils/dateUtils.js [hunk 29]
```diff
@@ -0,0 +1,37 @@
+/* eslint-disable require-jsdoc */
+
+
+export const getShortMonthFromString = (dateString) => {
+    try {
+        const parts = dateString.split('/');
+        const monthName = parts.length > 0 ? getMonth(new Date(parts[0], parts[1]-1), 'short') : ''
```
Comment (mareenanazir) [inline weatherman/utils/dateUtils.js]: what is this logic for?

File: weatherman/utils/mathUtils.js [hunk 30]
```diff
@@ -0,0 +1,23 @@
+/* eslint-disable require-jsdoc */
+
+export const getMaxObj = (objArray, property) => {
+    const data = removeEmptyValue(objArray, property)
+    return (data.length > 0) ? data.reduce((max, elem) => Number(max[property]) > Number(elem[property]) ? max : elem)  : data
+}
+
+export const getMinObj = (objArray, property) => {
+    const data = removeEmptyValue(objArray, property)
```
Comment (mareenanazir) [inline weatherman/utils/mathUtils.js]: why don't we clean the data before passing it to the function?

File: weatherman/weatherman.js [hunk 31]
```diff
@@ -0,0 +1,133 @@
...
+        else {
+            console.log("NO DATA FOUND FOR YEAR ", this.date)
+        }
+    }
+
+    /**
+     * Read Monthly weather Stats
+     */
+    calcualteMonthStats() {
+        this.withMonth = true
```
Comment (mareenanazir) [inline weatherman/weatherman.js]: name this to something better

File: weatherman/weatherman.js [hunk 32]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+     /**
+     * Claculate Highest Temperature, Lowest Temperature and
+     * Highest Humidity 
+     */
+    getHighLowTempAndHumidity() {
+        this.yearStats = {}
+        YEAR_WEATHER_STATS_COLUMNS.forEach(property => {
+            if (property != 'MinTemperatureC') {
```
Comment (mareenanazir) [inline weatherman/weatherman.js]: why dont we store this in a map to avoid if else

**Review by mareenanazir — CHANGES_REQUESTED** [review]

File: weatherman/index.js [hunk 33]
```diff
@@ -0,0 +1,57 @@
...
+  })
+  .help()
+  .alias("help", "h").argv;
+
+let directoryPath;
+
+if (argv.path) directoryPath = argv.path;
+else directoryPath = buildPath("weatherfiles");
+
+if (argv.calculateYearStats) {
```
Comment (mareenanazir) [inline weatherman/index.js:37]: what if the user does not press any of these options?

File: weatherman/utils/dateUtils.js [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+ * @param { string } monthFormat - format of return month "MMMMM" in case of
+ * full month name "MMM" in case of short name
+ */
+export const parseMonthArgsInputToMonth = (
+  inputDate,
+  inputFormat = "Y-M-D",
+  monthFormat = "MMMM"
+) => {
+  const parsedMonthDate = moment(inputDate, inputFormat, true);
+  return parsedMonthDate.isValid()
```
Comment (mareenanazir) [inline weatherman/utils/dateUtils.js]: is ternary operator required here? Or something else can be more appropriate here?

File: weatherman/utils/dateUtils.js [hunk 35]
```diff
@@ -0,0 +1,38 @@
...
+};
+
+/**
+ *
+ * @param { string } inputDate - input date
+ * @param { string } format - format of input date
+ */
+export const parseYearArgsInputToYear = (inputDate, format) => {
+  const parsedYearDate = moment(inputDate, format, true);
+  return parsedYearDate.isValid() ? String(parsedYearDate.year()) : undefined;
```
Comment (mareenanazir) [inline weatherman/utils/dateUtils.js]: can we replace ternary operator?

File: weatherman/utils/mathUtils.js [hunk 36]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+/**
+ *
+ * @param { Array } objArray - Array of objects
+ * @param { String } property - calculate min obj for this property
+ */
+export const getMinObj = (objArray, property) => {
+  return objArray.reduce((min, elem) =>
+    Number(min[property]) < Number(elem[property]) ? min : elem
```
Comment (mareenanazir) [inline weatherman/utils/mathUtils.js]: can't we do the typecasting while saving the data initially?

File: weatherman/weatherman.js [hunk 37]
```diff
@@ -0,0 +1,149 @@
...
+      console.log("INVALID DATE FORMAT", this.date);
+    }
+  }
+
+  /**
+   * Read Yearly weather Stats
+   */
+  calcualteYearStats() {
+    this.readFile(YEAR_WEATHER_STATS_COLUMNS);
+    if (this.weatherData.length > 0) {
```
Comment (mareenanazir) [inline weatherman/weatherman.js]: can we remove the if else?

**Review by mareenanazir — CHANGES_REQUESTED** [review]

**Pull request merged**